mutagen/dsf.py

```diff
@@ -0,0 +1,314 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (C) 2017  Boris Pruessmann
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of version 2 of the GNU General Public License as
+# published by the Free Software Foundation.
+
+"""Read and write DSF audio stream information and tags."""
+
+import sys
+import struct
+
+from ._compat import cBytesIO, reraise, endswith
+
+from mutagen import FileType, StreamInfo
+from mutagen._util import cdata, MutagenError, loadfile, convert_error
+from mutagen.id3 import ID3
+from mutagen.id3._util import ID3NoHeaderError, error as ID3Error
+
+
+__all__ = ["DSF", "Open", "delete"]
+
+
+class error(MutagenError):
+    pass
+
+
+class DSFHeaderError(error):
+    pass
+
+
+class DSFChunk(object):
+    """A generic chunk of a DSFFile."""
+
+    chunk_offset = 0
+    chunk_header = "    "
+    chunk_size = -1
+
+    def __init__(self, fileobj, create=False):
+        self.fileobj = fileobj
+
+        if not create:
+            self.chunk_offset = fileobj.tell()
+            self.load()
+
+    def load(self):
+        raise NotImplementedError
+
+    def write(self):
+        raise NotImplementedError
+
+
+class DSDChunk(DSFChunk):
+    """Represents the first chunk of a DSF file"""
+
+    CHUNK_SIZE = 28
+
+    total_size = 0
+    offset_metdata_chunk = 0
+
+    def __init__(self, fileobj, create=False):
+        super(DSDChunk, self).__init__(fileobj, create)
+
+        if create:
+            self.chunk_header = "DSD "
+            self.chunk_size = DSDChunk.CHUNK_SIZE
+
+    def load(self):
+        data = self.fileobj.read(DSDChunk.CHUNK_SIZE)
+        if not data.startswith(b"DSD "):
+            raise DSFHeaderError("DSF dsd header not found")
+
+        self.chunk_size = cdata.ulonglong_le(data[4:12])
+        if self.chunk_size != DSDChunk.CHUNK_SIZE:
+            raise DSFHeaderError("DSF dsd header size mismatch")
+
+        self.total_size = cdata.ulonglong_le(data[12:20])
+        self.offset_metdata_chunk = cdata.ulonglong_le(data[20:28])
+
+
+    def write(self):
+        f = cBytesIO()
+        f.write(b"DSD ")
+        f.write(struct.pack("<Q", DSDChunk.CHUNK_SIZE))
+        f.write(struct.pack("<Q", self.total_size))
+        f.write(struct.pack("<Q", self.offset_metdata_chunk))
+
+        self.fileobj.seek(self.chunk_offset)
+        self.fileobj.write(f.getvalue())
+
+    def pprint(self):
+        return u"DSD Chunk (Total file size = %d, Pointer to Metadata chunk = %d)" % (
+            self.total_size, self.offset_metdata_chunk)
+
+
+class FormatChunk(DSFChunk, StreamInfo):
```

Comment:
Please create a separate StreamInfo class (maybe taking the FormatChunk in __init__). This way the DSFChunk/FormatChunk API isn't exposed to the user.